File: 18-week-14/00-day-1-to-2--longest-substring-without-repeating-characters/solutions/longSubString.rb
def length_of_longest_substring(s)
    return s.size if s.size < 2
    
    queue, hash = [], {}
    s.each_char.reduce(-Float::INFINITY) do |max, c|
        hash.delete(char = queue.shift) until queue.empty? || char.eql?(c) if hash.key?(c)

        [max, (queue << hash[c] = c).size].max
    end
end


if __FILE__ == $PROGRAM_NAME

puts "Expecting: 3"
puts "=>", length_of_longest_substring("abcabcbb")

# Don't forget to add your own!
end
  

File: 18-week-14/00-day-1-to-2--longest-substring-without-repeating-characters/solutions/longSubString.js
function lengthOfLongestSubstring(s) {
    
  let map = {}
  let start = 0
  let maxLen = 0
  let arr = s.split('')

  for (i=0; i < s.length; i++) {
      let current = map[arr[i]]
      if (current!=null && start <= current) {
          start = current + 1
      } else {
          maxLen = Math.max(maxLen, i - start + 1)
      }
      
      map[arr[i]] = i
  }

  return maxLen
}

if (require.main === module) {

console.log("Expecting: 3");
console.log("=>", lengthOfLongestSubstring("abcabcbb"));

}

module.exports = lengthOfLongestSubstring